### PR TITLE
simplify anormal pass

### PR DESCRIPTION
**Pull request by amosr**
/jury approved @jystic

### CODE REVIEW

**Comment by jystic** [comment]
Seems reasonable, what's the story here?

**Comment by amosr** [comment]
the conversion to ANF is renaming way more variables than it needs to, and we should be able to take advantage of the fact that the input program never has shadowing, in order to limit renaming variables.

but this implementation is wrong, I just found a case which fails:

```
let a = (let b = 0 in b + 1)
\b. a + b

==>

let b = 0
let a = b + 1
\b. a + b
```

**Comment by amosr** [comment]
actually, it wasn't the renaming that was taking so long, it was the fact that the bookkeeping for what to rename was done in a separate pass, as well as converting a list to a set on every iteration.

**Comment by amosr** [comment]
A fact with 200 int fields
```
feature ints200 ~> group a1 ~> latest 1 ~> fields


45.2s -> 3.44s
```
!!!

**Comment by jystic** [comment]
> 45.2s -> 3.44s

Wow!!

**Comment by amosr** [comment]
yeah, it's crazy!

File: icicle-data/src/Icicle/Common/Exp/Simp/ANormal.hs [hunk 1]
```diff
@@ -59,65 +59,26 @@ anormal a_fresh xx
...
-  -- If the same name is used in multiple places, we'll introduce shadowing, which is
-  -- outlawed by the type checker.
-  --
-  -- So, we look at each binding, and if it would shadow something, we rename it.
-  renames _ [] seen  x
-   = return (seen, x)
-
-  renames allNames ((n,b):bs) seen x
-   -- Check if n is used anywhere else, including as a later binding name
-   |  n `Set.member` (allNames <> Set.fromList (fmap fst bs))
```
Comment (amosr) [inline icicle-data/src/Icicle/Common/Exp/Simp/ANormal.hs]: I think this was a large part of it - reconstructing this set every iteration, just to check membership. I wonder whether you could write a rewrite rule: `member n (a <> b) = member n a || member n b`

**Comment by jystic** [comment]
I can only imagine how bad it is for a 500 field struct

**Comment by amosr** [comment]
let's find out!

**Comment by amosr** [comment]
918.87s -> 31.56s

**Comment by tranma** [comment]
unreal

**Comment by jystic** [comment]
Winning, going to make a huge difference

**Comment by jystic** [comment]
:100:

**Pull request merged**